src/librustc_typeck/impl_wf_check.rs

```diff
@@ -104,35 +104,37 @@ fn enforce_impl_params_are_constrained<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     ctp::identify_constrained_type_params(
         tcx, &impl_predicates.predicates.as_slice(), impl_trait_ref, &mut input_parameters);
 
-    // Disallow ANY unconstrained type parameters.
-    for (ty_param, param) in impl_generics.types.iter().zip(impl_hir_generics.ty_params()) {
-        let param_ty = ty::ParamTy::for_def(ty_param);
-        if !input_parameters.contains(&ctp::Parameter::from(param_ty)) {
-            report_unused_parameter(tcx, param.span, "type", &param_ty.to_string());
-        }
-    }
-
     // Disallow unconstrained lifetimes, but only if they appear in assoc types.
     let lifetimes_in_associated_types: FxHashSet<_> = impl_item_refs.iter()
-        .map(|item_ref|  tcx.hir.local_def_id(item_ref.id.node_id))
+        .map(|item_ref| tcx.hir.local_def_id(item_ref.id.node_id))
         .filter(|&def_id| {
             let item = tcx.associated_item(def_id);
             item.kind == ty::AssociatedKind::Type && item.defaultness.has_value()
         })
         .flat_map(|def_id| {
             ctp::parameters_for(&tcx.type_of(def_id), true)
         }).collect();
-    for (ty_lifetime, lifetime) in impl_generics.regions.iter()
-        .zip(impl_hir_generics.lifetimes())
-    {
-        let param = ctp::Parameter::from(ty_lifetime.to_early_bound_region_data());
 
-        if
-            lifetimes_in_associated_types.contains(&param) && // (*)
-            !input_parameters.contains(&param)
-        {
-            report_unused_parameter(tcx, lifetime.lifetime.span,
-                                    "lifetime", &lifetime.lifetime.name.name().to_string());
+    for (ty_param, hir_param) in impl_generics.params.iter()
+                                              .zip(impl_hir_generics.params.iter()) {
+        match (&ty_param.kind, hir_param) {
+            // Disallow ANY unconstrained type parameters.
+            (&ty::GenericParamDefKind::Type(_), hir::GenericParam::Type(hir_ty)) => {
+                let param_ty = ty::ParamTy::for_def(ty_param);
+                if !input_parameters.contains(&ctp::Parameter::from(param_ty)) {
+                    report_unused_parameter(tcx, hir_ty.span, "type", &param_ty.to_string());
+                }
+            }
+            (&ty::GenericParamDefKind::Lifetime(_), hir::GenericParam::Lifetime(hir_lt)) => {
+                let param = ctp::Parameter::from(ty_param.to_early_bound_region_data());
+                if lifetimes_in_associated_types.contains(&param) && // (*)
+                    !input_parameters.contains(&param) {
+                    report_unused_parameter(tcx, hir_lt.lifetime.span,
+                                            "lifetime", &hir_lt.lifetime.name.name().to_string());
```

Comment:
I wonder if these could use `tcx.def_span(param.def_id)` and `param.name` to avoid looking at the HIR.